### PR TITLE
CAMEL-13429: Support expressions in path parameters at REST DSL

### PR DESCRIPTION
**Pull request by DenisIstomin**
Improved REST DSL route definition and REST Producer.
I have doubts about:
1. Using regex to iterate over placeholders "{}".
1. New function `resolveHeaderPlaceholders` at `camel-rest` that replaces placeholders with header values. Maybe there is a way to reuse some code from camel-properties component.

Also unit-tests for `camel-rest` component are located at `camel-core`. Maybe it would be better to move them?

### CODE REVIEW

**Comment by davsclaus** [comment]
Great work, and @bedlaj has some good observations.

And just to be sure the reg exp pattern is correct?

File: components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java [hunk 1]
```diff
@@ -223,6 +220,27 @@ protected void prepareExchange(Exchange exchange) throws Exception {
...
+     */
+    private String resolveHeaderPlaceholders(String a, Map<String, Object> headers) {
+        int startIndex = -1;
+        String res = a;
+        while ((startIndex = res.indexOf("{", startIndex + 1)) >= 0) {
+            int endIndex = res.indexOf("}", startIndex);
+            String key = res.substring(startIndex + 1, endIndex);
+            Object headerValue = headers.get(key);
+            if (headerValue != null) {
+                res = res.substring(0, startIndex) + headerValue.toString() + res.substring(endIndex + 1);
```
[the remark this answers is not in the document]
Comment (DenisIstomin) [reply components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java]: Thanks, made changes to use TypeConverters.

File: components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java [hunk 2]
```diff
@@ -223,6 +220,27 @@ protected void prepareExchange(Exchange exchange) throws Exception {
...
+     * @param a string with placeholders
+     * @param headers headers collection
+     * @return filled string
+     */
+    private String resolveHeaderPlaceholders(String a, Map<String, Object> headers) {
+        int startIndex = -1;
+        String res = a;
+        while ((startIndex = res.indexOf("{", startIndex + 1)) >= 0) {
+            int endIndex = res.indexOf("}", startIndex);
+            String key = res.substring(startIndex + 1, endIndex);
```
[the remark this answers is not in the document]
Comment (DenisIstomin) [reply components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java]: Thanks, added that check. Also added some unit-tests to cover that.

File: components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java [hunk 3]
```diff
@@ -155,18 +156,14 @@ protected void prepareExchange(Exchange exchange) throws Exception {
...
-                            hasPath = true;
-                            csb.append(value);
-                        } else {
-                            csb.append(a);
-                        }
-                    } else {
-                        csb.append(a);
+                    String message = resolveHeaderPlaceholders(a, inMessage.getHeaders());
+
+                    // If one of the headers is not resolved
```
[the remark this answers is not in the document]
Comment (DenisIstomin) [reply components/camel-rest/src/main/java/org/apache/camel/component/rest/RestProducer.java]: Thanks, made changes to code and that related comment.

**Comment by DenisIstomin** [comment]
@bedlaj, thanks for reviewing
Made changes and added some unit-tests.
@davsclaus to verify correctness of regex, improved unit-test `FromRestGetPlaceholderParamTest`.

**Comment by oscerd** [comment]
@DenisIstomin you can go ahead and merge

**Comment by DenisIstomin** [comment]
Thanks

**Pull request merged**